book/src/qobject/qobject_struct.md

```diff
@@ -5,42 +5,144 @@ SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
 SPDX-License-Identifier: MIT OR Apache-2.0
 -->
 
-# QObject Marked Struct
+# `#[cxx_qt::qobject]` Macro - Defining QObjects in Rust
+
+Defining QObjects is at the heart of CXX-Qt.
+Therefore `#[cxx_qt::qobject]` can be considered the most important macro in CXX-Qt.
+
+## Requirements
+- Like most other CXX-Qt macros, it can only be used from within a [`#[cxx_qt::bridge]`](./bridge-macro.md).
+- The `#[cxx_qt::qobject]` macro must be placed on a Rust struct.
+- The struct must `impl Default`, so that it can be constructed as part of a QObject.
+
+## Effects
+Adding the macro to a Rust struct `MyObject` has a few effects.
+
+However, first it's important to mention that this macro **does not modify the contents of the struct in any way!**
+This means you can always rely on the struct to behave like any other Rust struct.
+You can use it in normal Rust code, without interacting with Qt in any way.
+
+The macro does multiple other things for you though:
+- Generate a C++ QObject subclass that wraps the `MyObject` Rust struct.
+- Expose the generated QObject subclass to Rust as [`qobject::MyObject`](./generated-qobject.md)
+- Generate getters/setters for all fields.
+- Generate `Q_PROPERTY`s for all fields that are marked as `#[qproperty]`.
+- Generate signals if paired with a [`#[cxx_qt::qsignals]` enum](./signals_enum.md).
+
+### `base` attribute
+Use the `base` attribute to specify a C++ class that the C++ QObject will inherit from.
+The base class must inherit from QObject (directly or indirectly).
+
+``` rust,ignore,noplayground
+{{#include ../../../examples/qml_features/rust/src/custom_base_class.rs:book_qobject_base}}
+      // ...
+    }
+```
```

Comment:
Maybe out of scope for this PR, but I think it would be clearer to have an example for this that inherits from some Qt class rather than our own C++ class.

Reply:
Hopefully with 0.5 we can derive straight from a QAbstractListModel .. :-) If we can solve #330 (protected methods) #291 (QHash<K, V>).

We could change to another class but probably not worth it as hopefully this will change soon, as other classes may also need some of these improvements.